CMakeLists.txt

```diff
@@ -0,0 +1,630 @@
+
+include(CheckSymbolExists)
+include(CheckFunctionExists)
+include(CheckIncludeFile)
+include(TestBigEndian)
+
+# -----------------------------------
+# If platform is Emscripten
+if(${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten")
+    set(EMSCRIPTEN 1)
+endif()
+
+# Strip garbage
+if(APPLE)
+    string(REGEX REPLACE "-O3" ""
+        CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
+    set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2")
+    set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+
+    # Unify visibility to meet llvm's default.
+    include(CheckCCompilerFlag)
+    check_c_compiler_flag("-fvisibility-inlines-hidden" SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG)
+    if(SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG)
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fvisibility-inlines-hidden")
+    endif()
+elseif(NOT MSVC)
+    if(EMSCRIPTEN)
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Os -fdata-sections -ffunction-sections")
+        if("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+        else()
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+        endif()
+    else()
+        string(REGEX REPLACE "-O3" ""
+            CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
+        set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2 -fdata-sections -ffunction-sections")
+        if(ANDROID)
+            set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -funwind-tables")
+        elseif(NOT "${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
+            set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -s -Wl,--gc-sections -Wl,-s")
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -Wl,--gc-sections -Wl,-s")
+        else()
+            set(LINK_FLAGS_RELEASE  "${LINK_FLAGS_RELEASE} -dead_strip")
+        endif()
+    endif()
+endif()
+
+# Global optimization flags
+if(NOT MSVC)
+    set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -fno-omit-frame-pointer")
+endif()
+
+string(TOLOWER ${CMAKE_BUILD_TYPE} CMAKE_BUILD_TYPE_LOWER)
+
+if(CMAKE_BUILD_TYPE_LOWER STREQUAL "release")
+    add_definitions(-DNDEBUG)
+endif()
+
+if(MSVC)
+    # Force to always compile with W4
+    if(CMAKE_C_FLAGS MATCHES "/W[0-4]")
+        string(REGEX REPLACE "/W[0-4]" "/W4" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+    else()
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W4")
+    endif()
+elseif(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX)
+    # Update if necessary
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -pedantic")
+endif()
+
+# Disable bogus MSVC warnings
+if(MSVC)
+    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
+endif()
+
+# -fPIC thing
+if(NOT WIN32)
+    option(LIBXMP_PIC   "Build library as the position-independent code" OFF)
+    if(LIBXMP_PIC)
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
```

Comment:
Don't like this user-enabled/decided `-fPIC` flag: cmake already enables
it when building shared library if necessary. We don't need this fancy..

Reply:
I use it to build THIS library as static, and when I want to link it into another one which shared, mainly because in some cases I do making the monolithic shared library that links many static libraries inside of self.

Reply:
Otherwise, the link fails with the note to recompile that file with -fPIC.

Reply:
If you _truly_ want such an option, then confine it strictly to the static builds,
and name it accordingly like `build the **static** library as poisition independent`
or something.

Reply:
yeh, will do

Reply:
I'll also move the -fPIC into the static library target than global C_FLAGS